lib/csu/common-cheri/crt_init_globals.c

```diff
@@ -39,6 +39,47 @@
 #error "PIEs never need to initialise their own globals"
 #endif
 
+#ifdef CHERI_INIT_RELA
+extern const Elf_Rela __weak_symbol __rela_dyn_start __hidden;
+extern const Elf_Rela __weak_symbol __rela_dyn_end __hidden;
+
+static __always_inline void
+crt_init_rela(const Elf_Phdr *phdr __unused)
+{
+	const Elf_Rela *rela, *relalim;
+	void * __capability data_cap;
+	const void * __capability code_cap;
+
+#ifdef __CHERI_PURE_CAPABILITY__
+	data_cap = __DECONST(void *, phdr);
+#else
+	data_cap = cheri_getdefault();
+#endif
+	data_cap = cheri_clearperm(data_cap,
+	    CHERI_PERM_EXECUTE | CHERI_PERM_SW_VMEM);
+
+	code_cap = cheri_getpcc();
+
+#ifdef __CHERI_PURE_CAPABILITY__
+#if defined(__aarch64__)
+	__asm__ (
+	    "adrp %0, __rela_dyn_start\n\t"
```

Comment:
Can we put this in crt.h as some `#define FOO(x) ({ __typeof__(x) *_p; __asm__ ("adrp %0, %c1; add %0, %0, :lo12:%c1" : "=C"(_p) : "i"(&x)); _p; })`? Hybrid can define it to just `&x`.

Reply:
(Or crt1_c.c where CHERI_INIT_RELA lives)

Reply:
Then it can also be part of the second commit that adds Morello support specifically

Reply:
Oh, yeah, that might be fine, and certainly cleaner.  PCREL_PTR() maybe?

Reply:
Sure, seems fine, name doesn't hugely matter, it's internal to csu

Reply:
After adding a description in the commit log I ended up going with RODATA_PTR() for the name.